firedrake/mg/opencascade_mh.py

```diff
@@ -0,0 +1,250 @@
+from firedrake import *
+from .mesh import MeshHierarchy, HierarchyBase
+from .interface import prolong
+
+import numpy
+import subprocess
+import os
+import warnings
+
+
+__all__ = ("OpenCascadeMeshHierarchy",)
+
+
+def OpenCascadeMeshHierarchy(stepfile, element_size, levels, comm=COMM_WORLD, distribution_parameters=None, callbacks=None, order=1, mh_constructor=MeshHierarchy, cache=True, verbose=True, gmsh="gmsh", project_refinements_to_cad=True, reorder=None):
+
+    # OpenCascade doesn't give a nice error message if stepfile
+    # doesn't exist, it segfaults ...
+
+    try:
+        from OCC.Core.STEPControl import STEPControl_Reader
+        from OCC.Extend.TopologyUtils import TopologyExplorer
+    except ImportError:
+        raise ImportError("To use OpenCascadeMeshHierarchy, you must install firedrake with the OpenCascade python bindings (firedrake-update --opencascade).")
+
+    if not os.path.isfile(stepfile):
+        raise OSError("%s does not exist" % stepfile)
+
+    step_reader = STEPControl_Reader()
+    step_reader.ReadFile(stepfile)
+    step_reader.TransferRoot()
+    shape = step_reader.Shape()
+    cad = TopologyExplorer(shape)
+    dim = 3 if cad.number_of_solids() > 0 else 2
+    coarse = make_coarse_mesh(stepfile, cad, element_size, dim, comm=comm, distribution_parameters=distribution_parameters, cache=cache, verbose=verbose, gmsh=gmsh)
+
+    mh = mh_constructor(coarse, levels, distribution_parameters=distribution_parameters, callbacks=callbacks, reorder=reorder)
+    project_to_cad = project_mesh_to_cad_2d if dim == 2 else project_mesh_to_cad_3d
+
+    if project_refinements_to_cad:
+        for mesh in mh:
+            project_to_cad(mesh, cad)
+            push_coordinates_to_plex(mesh)
+        mh.nested = False
+
+    if order > 1:
+        VFS = VectorFunctionSpace
```

Comment:
Does this now work in parallel, or did the assertion of https://github.com/firedrakeproject/firedrake/pull/1448/commits/cfba05a86070faa78089e60e53bb434a209cfb97 get lost somewhere?

Reply:
Ah, I see that https://github.com/firedrakeproject/firedrake/pull/1448/commits/ee5f9eeea285e0468a8850f9d4fce469f8ff6226 fixes it?

Reply:
Yes, high-order construction now works in parallel and is tested in `tests/multigrid/test_opencascade_volume.py:test_area`.